From b3fbd7f7f82efef74566eac65202d53ffcbd9f03 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 16 Sep 2024 10:12:01 +0800
Subject: [PATCH 134/149] mtk: hostapd: skip BSS from disallowed channel when
 matching the scan results

In the single wiphy, we have added 'freq_list' in wpa_s configuration to
restrict legacy STA's scan & connection to specified band.
(For example, 2GHz legacy STA only scan 2GHz channels and should only
 connect to a 2GHz AP)

However, the legacy STA migth be configured on each band in a serious of
tests, and the test might also configure the AP with same SSID on each
band. As a result, the STA's scan results contain the BSS with same SSID
but from different bands, which might cause the STA to trigger association
from a different band than the user expected.

This commits skips the BSS from disallowed channel when the STA is
matching the scan results, ensuring that the connection can be triggered
on the correct band.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 wpa_supplicant/events.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/wpa_supplicant/events.c b/wpa_supplicant/events.c
index 435972eb2..8d86f2e77 100644
--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -1718,6 +1718,13 @@ struct wpa_ssid * wpa_scan_res_match(struct wpa_supplicant *wpa_s,
 		return NULL;
 	}
 
+	if (wpa_s->conf->freq_list &&
+	    !int_array_includes(wpa_s->conf->freq_list, bss->freq)) {
+		if (debug_print)
+			wpa_dbg(wpa_s, MSG_DEBUG, "  skip - channel disallowed");
+		return NULL;
+	}
+
 	if (wnm_is_bss_excluded(wpa_s, bss)) {
 		if (debug_print)
 			wpa_dbg(wpa_s, MSG_DEBUG, "   skip - BSSID excluded");
-- 
2.45.2

